sdk/resourcemanager/azure-resourcemanager-resources/src/main/java/com/azure/resourcemanager/resources/fluentcore/arm/ResourceUtils.java

```diff
@@ -156,23 +156,28 @@ public static String nameFromResourceId(String id) {
      * @return the default api version to use
      */
     public static String defaultApiVersion(String id, Provider provider) {
-        String resourceType = resourceTypeFromResourceId(id).toLowerCase(Locale.ROOT);
+        ResourceId resourceId = ResourceId.fromString(id);
+        String resourceType = resourceId.resourceType().toLowerCase(Locale.ROOT);
         // Exact match
         for (ProviderResourceType prt : provider.resourceTypes()) {
             if (prt.resourceType().equalsIgnoreCase(resourceType)) {
                 return prt.defaultApiVersion() == null ? prt.apiVersions().get(0) : prt.defaultApiVersion();
             }
         }
-        // child resource, e.g. sites/config
-        for (ProviderResourceType prt : provider.resourceTypes()) {
-            if (prt.resourceType().toLowerCase(Locale.ROOT).contains("/" + resourceType)) {
-                return prt.defaultApiVersion() == null ? prt.apiVersions().get(0) : prt.defaultApiVersion();
+        // child resource, e.g. Microsoft.Web/sites/config
+        // There is an edge case that two resource types share the same child resource type name, so parent type name check is needed.
+        // e.g. "dnsForwardingRulesets/virtualNetworkLinks" and "privateDnsZones/virtualNetworkLinks"
+        ResourceId parent = resourceId.parent();
+        if (parent != null) {
+            for (ProviderResourceType prt : provider.resourceTypes()) {
+                String fullResourceType = resourceId.fullResourceType().toLowerCase(Locale.ROOT);
+                // some resource types are full resource type, e.g. Microsoft.Web/sites/config
+                // some are not, e.g. privateDnsZones/virtualNetworkLinks
+                if (fullResourceType.endsWith(prt.resourceType().toLowerCase(Locale.ROOT))) {
```

Comment:
Is `prt.resourceType()` always full resource type (include namespace)?

Reply:
Double checked, actually no.. The `prt.resourceTypes()` doesn't contain the namespace. It's the response from the WebApp client API that does, `Microsoft.Web/sites/config`. 
I wasn't able to find whether it ever will from their api docs: https://learn.microsoft.com/en-us/rest/api/resources/provider-resource-types/list?tabs=HTTP#providerresourcetypelistresult . 
But I assume they won't since their examples don't: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-providers-and-types

I'll change the implementation to compare the exact resource type without namespace.

And `prt.resourceTypes()` doesn't contain `sites/confg` either, which always makes the test case use the default api-version.. I'll delete the test for this one.

Reply:
Wondering whether we should make 2 pass. 1st to check for exact match of 2 resourceType. If 1st pass not found, do a 2nd pass with relaxed condition.

Reply:
By `current condition` you mean the full resource type right? Sounds good to me.

Reply:
By "current condition" I mean the `endsWith` (or maybe "previous condition" of `contains`?). Point is a relaxed match.

Reply:
updated